elastic/kibana
# Onboard alerting tasks to use stateSchemaByVersion for task state validation of the framework fields

Resolves https://github.com/elastic/kibana/issues/159343

In this PR, I'm preparing the alerting rule task types for serverless by defining an explicit task state schema for the framework level fields. This schema is used to validate the task's state before saving but also when reading. In the scenario an older Kibana node runs a task after a newer Kibana node has stored additional task state, the unknown state properties will be dropped. Additionally, this will prompt developers to be aware that adding required fields to the task state is a breaking change that must be handled with care. (see https://github.com/elastic/kibana/issues/155764).

The PR also includes the following changes:
- Modifying the `@kbn/alerting-state-types` package so the types are generated from `config-schema`, instead of `io-ts`. The schema is re-used for exporting a new `stateSchemaByVersion` property.
- Removing `DateFromString` in favour of strings everywhere (config-schema doesn't support this conversion)
- Add a v1 `up` migration that will ensure the types match the TypeScript interface on any existing alerting task. The migration assumes any type of data could exist and dropping parts that don't match the type expectation. The TypeScript interface uses `schema.maybe()` in a lot of places (as `io-ts` did), so safe if ever data gets dropped.
- Cleanup the `alerting/common/**` exports to reduce bundle size. Because the `@kbn/alerting-state-types` package grew.
- Since the new TypeScript interfaces / types are `ReadOnly<...>`, I created some `Mutable...` types for places that needed it (in order to avoid code refactoring).

## To verify

Stack Monitoring:
- To make TypeScript happy with the new ReadOnly `RawAlertInstance` type, I removed some redundant code and solved the issue.

Security Solution:
- Changes to the `alertInstanceFactoryStub` set the alert's date to a `string` instead of a `Date` value. Note: The HTTP API response converted `Date` objects to `string`, so the HTTP API response will look the same with this change.

Response Ops:
- In a fresh Kibana install, create alerting rules and ensure they run.
- In a 8.9 version, create some rules, upgrade to this branch and ensure they still run.
- Compare the `io-ts` definition with the new `config-schema`. They should match 1:1.
- Look for ways the migration code could fail.

Note: The main changes are within the following areas:
- `x-pack/plugins/alerting/server/rule_type_registry.ts`
- `x-pack/packages/kbn-alerting-state-types/*`
